refactor (shell) : detect shell on windows using gopsutil too (#4588)

rohanKanojia opened the pull request:
## Description

<!--
Thank you for your pull request (PR)!

Please provide a description of what your PR does providing a link (if applicable) to the issue it fixes.
-->
As a follow up to #4572, updating shell detection logic on windows to also rely on gopsutil library to detect the shell name by inspecting parent processes of `crc.exe` process.

Fixes: #4588

Relates to: PR #4572

<!--
Describe in plain English what you solved and how. For instance, _Added `start` command to CRC so the user can create a VM and set up a single-node OpenShift cluster on it with one command. It requires blablabla..._
-->

## Type of change
<!--
What types of changes does your code introduce? Put an `x` in all the boxes that apply
-->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change
- [x] Chore (non-breaking change which doesn't affect codebase;
  test, version modification, documentation, etc.)

## Proposed changes
<!--
List main as well as consequential changes you introduced or had to introduce.

1. Add `start` command.
2. Add `setup` as prerequisite to `start`.
3. ...
-->
This PR mostly includes moving some methods so that they become accessible in windows environments too.

+ Move shell detection using `gopsutil` logic from `shell_unix.go` -> `shell.go` so that it can be reused in `shell_windows.go` as well
+ Remove `shell_linux.go`/ `shell_darwin.go` as they only contained a single constant that I moved to `shell_unix.go`
+ Remove `getProcessEntry` and related methods from `shell_windows.go` as we're reusing gopsutil approach that is working well on Windows

## Testing
<!--
What is the _bottom-line_ functionality that needs testing? Describe in pseudo-code or in English. Use verifiable statements that tie your changes to existing functionality.

1. `start` succeeds first time after `setup` succeeded
2. stdout contains ... if `start` succeeded
3. stderr contains ... after `start` failed
4. `status` returns ... if `start` succeeded
5. `status` returns ... if `start` failed
6. `start` fails after `start` succeeded or after `status` says CRC is `Running`
7. ...
-->
1. Start CRC in windows environment
2. Verify `crc oc-env` is working as expected in 
    - CMD
    - PowerShell
    - GitBash
    - WSL
      - bash
      - fish
      - zsh

## Contribution Checklist
- [x] I Keep It Small and Simple: The smaller the PR is, the easier it is to review and have it merged
- [ ] I have performed a self-review of my code
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Which platform have you tested the code changes on? <!-- Only put an `x` in applicable platforms -->
    - [ ] Linux
    - [x] Windows
    - [x] MacOS

cfergeau commented on pkg/os/util.go:
```diff
@@ -127,3 +127,12 @@ func RemoveFileGlob(glob string) error {
 	return nil
 }
+
+func IsPresentInList(arr []string, str string) bool {
```
It already exists as [Contains](https://github.com/crc-org/crc/blob/main/pkg/strings/strings.go#L8)

rohanKanojia replied:
Thanks, I was going to use this but I noticed a unit test is failing on Windows due to these changes. This line was returning bash on Windows powershells:

https://github.com/crc-org/crc/blob/6d72284a886c4978e13e4775b9b8f552e134bf18/cmd/crc/cmd/start_test.go#L41

On checking the process tree during unit test execution, it shows processes like this:
`cmd.test.exe`  -> `go.exe`     -> `bash.exe`

I've added another utility method in order to just match the prefix instead of matching entire string.

cfergeau commented on pkg/os/shell/shell.go:
```diff
@@ -151,3 +171,50 @@ func IsWindowsSubsystemLinux() bool {
+		if err != nil {
+			return ""
+		}
+		if crcos.IsPresentInList(supportedShell, processName) {
```
This would be easier to review if it did not mix code movement with functional changes. I'd suggest having a commit where this code is moved to `shell.go` without any changes, and  then a commit porting the Windows code to use `detectShellByCheckingProcessTree` with these `crcos.IsPresentInList` and such.

rohanKanojia replied:
Thanks, I've split the PR into two commits with first commit containing code reorganization and second including actual change

rohanKanojia committed:
refactor (shell) : Prerequisite code reorganization for gopsutil windows refactor

Related to #4588

Move shell detection related methods from shell_unix.go to shell.go so
that they can be accessed from Windows code as well

Signed-off-by: Rohan Kumar <rohaan@redhat.com>

rohanKanojia committed:
refactor (shell) : Windows Shell detection uses gopsutil (#4588)

+ Update `shell_windows.go` to use detectShellByCheckingProcessTree
instead of relying on SHELL environment variable.
+ Remove hardcoded check from detectShellByCheckingProcessTree for shell
  types, use already present supportedShell slice.

Signed-off-by: Rohan Kumar <rohaan@redhat.com>

cfergeau commented on pkg/os/shell/shell.go:
```diff
@@ -208,7 +210,9 @@ func detectShellByCheckingProcessTree(p AbstractProcess) string {
-		if processName == "zsh" || processName == "bash" || processName == "fish" {
+		if crcstrings.IsPresentInListSatisfying(supportedShell, processName, func(listElem string, toMatch string) bool {
+			return strings.HasPrefix(toMatch, listElem)
+		}) {
```
note: if you switch the  2 arguments when invoking the callback in `IsPresentInListSatisfying`, I think this can be simplified to `crcstrings.IsPresentInListSatisfying(supportedShell, processName, strings.HasPrefix)`. Fine with me if you prefer to keep the code the way it is now.

cfergeau commented on pkg/strings/strings.go:
```diff
@@ -6,8 +6,14 @@ import (
+	})
+}
+
+func IsPresentInListSatisfying(input []string, toMatch string, matchingPredicate func(string, string) bool) bool {
```
While looking for naming ideas in go standard packages (usual name for such a function is `ContainsFunc`), I realized the [`slices`](https://pkg.go.dev/slices#ContainsFunc) package has both `Contains` and `ContainsFunc`, we can use these instead.

rohanKanojia replied:
Oh, thanks a lot! I'm sorry I should've checked this.

cfergeau reviewed (approved):
There's an outdated `+ Add a utility method in strings for verifying a slice contains an
  element matching the provided predicate.` in the commit log of the second commit, otherwise looks good to me!

rohanKanojia commented:
@cfergeau : Thanks for noticing! I've updated the commit message.

I'll also create a follow up PR to make use of [slice.Contains](url) to replace the equivalent method in crcstrings.Contains
